### PR TITLE
Added Resource Usage section to Resource Quota Details page

### PR DESCRIPTION
**Pull request by dtaylor113**
**Administration -> Resource Quotas -> Click on a quota, at bottom of details page:**
image

image

image

JIRA: https://jira.coreos.com/browse/CONSOLE-685
* PR does not include guage/donut charts, these will be done in a follow up PR

### CODE REVIEW

File: frontend/public/components/resource-quota.jsx [hunk 1]
```diff
@@ -13,6 +13,100 @@ const menuActions = [...common];
...
+    'requests.cpu': 'CPU (Request)',
+    'requests.memory': 'Memory (Request)',
+    'resourcequotas': 'Resource Quotas',
+    'secrets': 'Secrets',
+    'services': 'Services',
+    'services.loadbalancers': 'Service Load Balancers',
+    'services.nodeports': 'Service Node Ports',
+  };
+
+  const nameFormatMap = {
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]: Better to have one map that has both lowercase and title case values. That way we can make sure the maps have all the same keys.

```js
const quotaResourceLabels = Object.freeze({
  configmaps: { label: 'config maps', titleCase: 'Config Maps' },
  // ...
});

const humanizeQuotaResource = (resourceType, useTitleCase) =>
  _.get(quotaResourceLabels, [resourceType, useTitleCase ? 'titleCase' : 'label'], resourceType);
```

File: frontend/public/components/_quota.scss [hunk 2]
```diff
@@ -0,0 +1,7 @@
+.co-m-resource-quota-icon {
+  padding-left: 34px;
+}
+
+.co-m-resource-quota-empty {
+  font-size: 15px;
```
Comment (spadgett) [inline frontend/public/components/_quota.scss]: Curious why we need a special rule here.

File: frontend/public/components/_quota.scss [hunk 3]
```diff
@@ -0,0 +1,7 @@
+.co-m-resource-quota-icon {
```
Comment (spadgett) [inline frontend/public/components/_quota.scss]: We should drop the `-m-`.

```suggestion
.co-resource-quota-icon {
```

File: frontend/public/components/_quota.scss [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+  padding-left: 34px;
+}
+
+.co-m-resource-quota-empty {
```
Comment (spadgett) [inline frontend/public/components/_quota.scss]:
```suggestion
.co-resource-quota-empty {
```

File: frontend/public/components/resource-quota.jsx [hunk 5]
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
...
+  } else if (capacity > 0 && capacity < .5) {
+    capacityIconClass = 'pficon-resources-almost-empty';
+  } else if (capacity >= .5 && capacity < 1){
+    capacityIconClass = 'pficon-resources-almost-full';
+  } else if (capacity === 1) {
+    capacityIconClass = 'pficon-resources-full';
+  } else if (capacity > 1) {
+    capacityIconClass = 'pficon-warning-triangle-o';
+  }
+  const classPrefix = capacityIconClass.startsWith('pficon') ? 'pficon' : 'fa';
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]: Let's just add `fa` or `pficon` to the strings above rather than trying to calculate a prefix. The icons are all known, and it's less error prone.

File: frontend/public/components/resource-quota.jsx [hunk 6]
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
...
+  return <i className={`${classPrefix} ${capacityIconClass}`} aria-hidden="true" />;
+};
+
+const getCapacity = (used, max, resourceType) => {
+  let capacity = 0;
+  if (max && used) {
+    let unitType = units.getUnitType(max);
+    if (!unitType) {
+      unitType = units.getUnitType(used);
+    }
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]:
```js
const unitType = units.getUnitType(max) || units.getUnitType(used);
```

File: frontend/public/components/resource-quota.jsx [hunk 7]
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
...
+  } else if (capacity === 1) {
+    capacityIconClass = 'pficon-resources-full';
+  } else if (capacity > 1) {
+    capacityIconClass = 'pficon-warning-triangle-o';
+  }
+  const classPrefix = capacityIconClass.startsWith('pficon') ? 'pficon' : 'fa';
+  return <i className={`${classPrefix} ${capacityIconClass}`} aria-hidden="true" />;
+};
+
+const getCapacity = (used, max, resourceType) => {
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]: This seems misnamed. Capacity usually means the total amount available. Maybe `getDecimalUsage` here and `UsageIcon` above?

File: frontend/public/components/resource-quota.jsx [hunk 8]
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
...
+  let capacity = 0;
+  if (max && used) {
+    let unitType = units.getUnitType(max);
+    if (!unitType) {
+      unitType = units.getUnitType(used);
+    }
+    // max cpu values do not have units, convert number and add unit for capacity calculation
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]: This is wrong. Max CPU values CAN have a unit that we need to account for. We need to make sure we're using the utilties from units for any compute resource.

File: frontend/public/components/resource-quota.jsx [hunk 9]
```diff
@@ -13,6 +13,100 @@ const menuActions = [...common];
...
+    return 1;
+  }
+
+  left = humanizeQuotaResource(left);
+  right = humanizeQuotaResource(right);
+  return left.localeCompare(right);
+};
+
+const getResourceTypes = (quota) => {
+  const specHard = _.get(quota, 'spec.quota.hard') || _.get(quota, 'spec.hard');
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]: Are you trying to handle ClusterResourceQuota? Let's do that in a separate PR.

File: frontend/public/components/resource-quota.jsx [hunk 10]
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
...
+    }
+    maxNum = units.dehumanize(maxStr, unitType).value;
+    const usedNum = units.dehumanize(used, unitType).value;
+    capacity = usedNum / maxNum;
+  }
+  return capacity;
+};
+
+const ResourceUsageRow = ({quota, resourceType}) => {
+  const max = quota.status.hard ? quota.status.hard[resourceType] : quota.spec.hard[resourceType];
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]:
```suggestion
  const max = _.get(quota, ['status', 'hard', resourceType]) || _.get(quota, ['spec', 'hard', resourceType]);
```

(And similar for below.)

File: frontend/public/components/utils/units.js [hunk 11]
```diff
@@ -147,6 +147,78 @@ units.dehumanize = (value, typeName) => {
   return convertValueWithUnitsToBaseValue(value, type.units, type.divisor);
 };
 
+const humanizeUsageUnit = (unit, type, singular) => {
```
Comment (spadgett) [inline frontend/public/components/utils/units.js]: Why can't we use the existing humanize methods? There is already `units.humanize`

File: frontend/public/components/utils/units.js [hunk 12]
```diff
@@ -147,6 +147,78 @@ units.dehumanize = (value, typeName) => {
...
+      }
+      const suffix = (singular) ? 'core' : 'cores';
+      return (unit || '') + suffix;
+    }
+    default:
+      return unit;
+  }
+};
+
+const usageAmountAndUnit = (value, type, humanizeUnits) => {
```
Comment (spadgett) [inline frontend/public/components/utils/units.js]: You shouldn't need this. Use `units.dehumanize`.

File: frontend/public/components/resource-quota.jsx [hunk 13]
```diff
@@ -30,11 +124,77 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
...
+    <div className="row">
+      <div className="co-m-table-grid co-m-table-grid--bordered">
+        <div className="row co-m-table-grid__head">
+          <div className="col-xs-4">Resource Type</div>
+          <div className="col-xs-2">Capacity</div>
+          <div className="col-xs-3">Used</div>
+          <div className="col-xs-3">Max</div>
+        </div>
+        <div className="co-m-table-grid__body">
+          {getResourceTypes(rq).map((t, i) => <ResourceUsageRow key={i} quota={rq} resourceType={t} />)}
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]:
```suggestion
          {getResourceTypes(rq).map(type => <ResourceUsageRow key={type} quota={rq} resourceType={type} />)}
```

**Comment by spadgett** [comment]
@cshinn @dtaylor113 I like the way this is presented 👍

File: frontend/public/components/resource-quota.jsx [hunk 14]
```diff
@@ -13,6 +13,100 @@ const menuActions = [...common];
 
 const quotaKind = quota => quota.metadata.namespace ? referenceForModel(ResourceQuotaModel) : referenceForModel(ClusterResourceQuotaModel);
 
+const humanizeQuotaResource = (resourceType, useTitleCase) => {
+  if (!resourceType) {
+    return resourceType;
+  }
+
+  const nameTitleCaseFormatMap = {
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: I see usage like:
```
import { Map as ImmutableMap } from 'immutable';
export const resourceDetailPages = ImmutableMap<GroupVersionKind | string, () => Promise<React.ComponentType<any>>>()
  .set(...
```
However, I'm in a `.jsx` file whereas 'immutable' seems to be a Typescript library.  Ok if I use `let map = new Map();`  I don't think it's immutable though.

File: frontend/public/components/utils/units.js (hunk 11 above)
Comment (spadgett) [reply frontend/public/components/utils/units.js]: Actually, I'd suggest we should just show the string value as-is since that's what we do in other places in the console. Then we can use `units.dehumanize` for comparing values and calculating percent usage.

File: frontend/public/components/resource-quota.jsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (spadgett) [reply frontend/public/components/resource-quota.jsx]: You shouldn't need TypeScript.

https://facebook.github.io/immutable-js/docs/#/

File: frontend/public/components/utils/units.js (hunk 11 above)
Comment (spadgett) [reply frontend/public/components/utils/units.js]: > Actually, I'd suggest we should just show the string value as-is since that's what we do in other places in the console.

For example,

okd 2018-11-16 14-02-38

We should show these the same way here to avoid confusion.

File: frontend/public/components/resource-quota.jsx [hunk 15]
```diff
@@ -13,6 +13,100 @@ const menuActions = [...common];
...
+  // CPU Request
+  if (left === 'cpu' || left === 'requests.cpu') {
+    return right === 'cpu' || right === 'requests.cpu' ? 0 : -1;
+  }
+  if (right === 'cpu' || right === 'requests.cpu') {
+    return 1;
+  }
+
+  // Memory Request
+  if (left === 'memory' || left === 'requests.memory') {
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]: I'd suggest we order it like

CPU (requests)
CPU (limits)
Memory (requests)
Memory (limits)

File: frontend/public/components/_quota.scss (hunk 2 above)
Comment (dtaylor113) [reply frontend/public/components/_quota.scss]: Because we are mixing `fa` and `pf` icons, this is needed so that `fa` icons are center aligned with the `pf` icons

File: frontend/public/components/utils/units.js (hunk 11 above)
Comment (dtaylor113) [reply frontend/public/components/utils/units.js]: @spadgett, ok with showing units 'as-is', but wondering about the CPU units of 'cores'.  Max CPU only shows an integer (ex. '1') should this be converted to `1000 m'? -thanks

File: frontend/public/components/resource-quota.jsx (hunk 5 above)
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: I can't win :-) I initially had it that way, but thought you'd prefer the more concise logic :-)

File: frontend/public/components/_quota.scss (hunk 2 above)
Comment (spadgett) [reply frontend/public/components/_quota.scss]: It seems like there should be a `pficon-resources-empty` so we don't have to mix in the font awesome icon for this one special case.

FYI @serenamarie125 there are

* `pficon-resources-almost-empty`
* `pficon-resources-almost-full`
* `pficon-resources-full`

But no pficon for empty

File: frontend/public/components/utils/units.js (hunk 11 above)
Comment (spadgett) [reply frontend/public/components/utils/units.js]: I would leave it as-is for now at least.

File: frontend/public/components/resource-quota.jsx (hunk 1 above)
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: fixed

File: frontend/public/components/_quota.scss (hunk 3 above)
Comment (dtaylor113) [reply frontend/public/components/_quota.scss]: fixed

File: frontend/public/components/_quota.scss (hunk 4 above)
Comment (dtaylor113) [reply frontend/public/components/_quota.scss]: fixed

File: frontend/public/components/resource-quota.jsx (hunk 5 above)
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: fixed

File: frontend/public/components/resource-quota.jsx (hunk 9 above)
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: fixed

File: frontend/public/components/resource-quota.jsx (hunk 10 above)
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: fixed

File: frontend/public/components/utils/units.js (hunk 11 above)
Comment (dtaylor113) [reply frontend/public/components/utils/units.js]: fixed

File: frontend/public/components/resource-quota.jsx (hunk 13 above)
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: fixed

File: frontend/public/components/utils/units.js (hunk 12 above)
Comment (dtaylor113) [reply frontend/public/components/utils/units.js]: fixed

File: frontend/public/components/resource-quota.jsx (hunk 8 above)
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: If there is a unit for max cpu it will be used as is, this only handles when there is not a unit.

Comment (spadgett) [reply frontend/public/components/resource-quota.jsx]: But min could also not have a unit. There's nothing special about max CPU. It's also possible to specify memory values in bytes with no units.

File: frontend/public/components/resource-quota.jsx (hunk 6 above)
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: fixed

File: frontend/public/components/resource-quota.jsx (hunk 8 above)
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: I am seeing examples where min has a unit, but max does not:

image
image

Comment (spadgett) [reply frontend/public/components/resource-quota.jsx]: Right, but that's just one specific example. It's not required for any CPU or memory value. `requests.cpu` could be "1" and `limits.cpu` could be 1500m.

You should just be able to always call `dehumanize` and work with the number that comes back.

Comment (spadgett) [reply frontend/public/components/resource-quota.jsx]: > Well, the `200m` will result in a unitType of 'numeric', but calling `units.dehumanize('1', 'numeric')` results in a 0 -the logic in dehumanize doesn't handle this case.

You need to pass the CPU type for it to handle it correctly. It's not numeric. It's CPU.

I don't think you can infer the unit type. You need to specifically pass CPU for CPU, memory for memory, etc.

File: frontend/public/components/resource-quota.jsx (hunk 7 above)
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: fixed

File: frontend/public/components/resource-quota.jsx (hunk 8 above)
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: I do not see a CPU type in  https://github.com/openshift/console/blob/master/frontend/public/components/utils/units.js#L6  ?

Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: The two CPU dehumanize examples I see are using 'numeric': https://github.com/openshift/console/blob/master/frontend/public/components/pod.jsx#L161
https://github.com/openshift/console/blob/master/frontend/public/components/pod.jsx#L165

Comment (spadgett) [reply frontend/public/components/resource-quota.jsx]: > The two CPU dehumanize examples I see are using 'numeric': 

OK, I see now. I think that's just a bug. We need to add correct handling for CPU units.

Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: > OK, I see now. I think that's just a bug. We need to add correct handling for CPU units.

Ok, added `dehumanizeCPU` method

File: frontend/public/components/resource-quota.jsx [hunk 16]
```diff
@@ -85,16 +85,15 @@ const UsageIcon = ({capacity}) => {
...
   if (max && used) {
-    const unitType = units.getUnitType(max) || units.getUnitType(used);
-    // max cpu values do not have units, convert number and add unit for capacity calculation
-    // Ex. converts '1' to '1000m'
-    let maxStr = max;
-    let maxNum = _.toNumber(max);
-    if (_.includes(resourceType, '.cpu') && unitType && !_.isNaN(maxNum)) { // skip if max cpu already has a unit
-      maxStr = `${maxNum * 1000}m`;
+    let usedNum = 0, maxNum = 0;
+    if (_.includes(resourceType, '.cpu')) {
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]: I'm a little uneasy about the `getUnitType` detection. I'm not sure we can reliably detect.

I suggest creating sets of known CPU and memory/size quota resources and calling the right dehumanize method for those.

File: frontend/public/components/utils/units.js [hunk 17]
```diff
@@ -138,15 +138,58 @@ const humanize = units.humanize = (value, typeName, useRound = false) => {
 };
 
 export const humanizeMem = v => humanize(v, 'binaryBytes', true).string;
-export const humanizeCPU = v => humanize(v, 'numeric', true).string;
 export const humanizeNumber = v => humanize(v, 'numeric', true).string;
+export const humanizeCPU = v => humanize(v, 'numeric', true).string;
+export const dehumanizeCPU = v => {
+  const _dehumanizeCPU = _.spread((value, unit) => {
```
Comment (spadgett) [inline frontend/public/components/utils/units.js]: Prefer es6 syntax to `_.spread`:

```suggestion
  const _dehumanizeCPU = ([value, unit]) => {
```

File: frontend/public/components/utils/units.js [hunk 18]
```diff
@@ -138,15 +138,58 @@ const humanize = units.humanize = (value, typeName, useRound = false) => {
...
+    if (!unit) {
+      return {value: value * 1000, unit: 'm'};
+    }
+    return {value, unit};
```
Comment (spadgett) [inline frontend/public/components/utils/units.js]: This is not guaranteed to return values that always have the same unit.

Comment (dtaylor113) [reply frontend/public/components/utils/units.js]: for CPU isn't it 'm' or nothing?  What other units could there be?

**Comment by dtaylor113** [comment]
> The way you've written this, I think we can reuse the existing dehumanize logic with type:
> 
> ```js
> cpu: {
>   units: ['m', ''],
>   space: false,
>   divisor: 1000,
> },
> ```
> 
> Can you try this first?

@spadgett 

const test = units.dehumanize('1', 'cpu').value;     // returns "1"
const test2 = units.dehumanize('100m', 'cpu').value;  // returns "100m"

* trying other values in units array, so far no luck

File: frontend/public/components/resource-quota.jsx (hunk 16 above)
Comment (spadgett) [reply frontend/public/components/resource-quota.jsx]: You should be able to use #852 to get the base values now without needing to check unit types.

File: frontend/public/components/resource-quota.jsx [hunk 19]
```diff
@@ -13,6 +13,42 @@ const menuActions = [...common];
...
+  'requests.storage': {label: 'storage request', titleCase: 'Storage (Request)'},
+  'pods': {label: 'pods', titleCase: 'Pods'},
+  'replicationcontrollers': {label: 'replication controllers', titleCase: 'Replication Controllers'},
+  'requests.cpu': {label: 'cpu request', titleCase: 'CPU (Request)'},
+  'requests.memory': {label: 'memory request', titleCase: 'Memory (Request)'},
+  'resourcequotas': {label: 'resource quotas', titleCase: 'Resource Quotas'},
+  'secrets': {label: 'secrets', titleCase: 'Secrets'},
+  'services': {label: 'services', titleCase: 'Services'},
+  'services.loadbalancers': {label: 'service load balancers', titleCase: 'Service Load Balancers'},
+  'services.nodeports': {label: 'service node ports', titleCase: 'Service Node Ports'},
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]: @dtaylor113 This looks like an incomplete list.

https://docs.okd.io/latest/admin_guide/quota.html#managed-by-quota

I'm worried that this is constantly going to get out of date, so we have a mix of humanized / unhumanized labels. We might skip this and just show the real value. It's also confusing to see something different in the table than you need to add the to YAML.

cc @cshinn

Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: @spadgett, @cshinn, so instead of:
image
We'd show:
```
    limits.cpu
    requests.cpu
    limits.memory
    requests.memory
    pods
```

**Comment by dtaylor113** [comment]
Hi @spadgett, updated to use new 'convertToBaseValue` function.

File: frontend/public/components/resource-quota.jsx (hunk 19 above)
Comment (spadgett) [reply frontend/public/components/resource-quota.jsx]: @dtaylor113 I think that's better because we're never going to be able to humanize everything, and things will always be added later. Storage classes follow the format `<storage-class-name>.storageclass.storage.k8s.io/requests.storage`, so those will pretty much always have the raw name unless we do some regex matching or something. Plus it doesn't match what's in the YAML or CLI or docs, so it's harder to know what to change if you want to edit something. We had a number of bugs around trying to humanize quota keys in the previous console.

I think this is a losing battle, and we should just show the quota key as-is.

Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: Ok, thanks, sounds reasonable to me.  @cshinn, sound good to you?  -thanks

Comment (cshinn) [reply frontend/public/components/resource-quota.jsx]: yeah, that works for me. Especially if we're likely to have other things added

File: frontend/public/components/resource-quota.jsx [hunk 20]
```diff
@@ -30,11 +66,65 @@ const Row = ({obj: rq}) => <div className="row co-resource-list__item">
...
+  } else if (capacity > 0 && capacity < .5) {
+    capacityIconClass = 'pficon pficon-resources-almost-empty';
+  } else if (capacity >= .5 && capacity < 1){
+    capacityIconClass = 'pficon pficon-resources-almost-full';
+  } else if (capacity === 1) {
+    capacityIconClass = 'pficon pficon-resources-full';
+  } else if (capacity > 1) {
+    capacityIconClass = 'pficon pficon-warning-triangle-o';
+  }
+  return <i className={`${capacityIconClass}`} aria-hidden="true" />;
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: fixed

File: frontend/public/components/resource-quota.jsx [hunk 21]
```diff
@@ -13,6 +13,42 @@ const menuActions = [...common];
...
+  'services.nodeports': {label: 'service node ports', titleCase: 'Service Node Ports'},
+});
+
+const humanizeQuotaResource = (resourceType, useTitleCase) =>
+  _.get(quotaResourceLabels, [resourceType, useTitleCase ? 'titleCase' : 'label'], resourceType);
+
+const compareResourceType = (left, right) => {
+  left = humanizeQuotaResource(left);
+  right = humanizeQuotaResource(right);
+  return left.localeCompare(right);
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/resource-quota.jsx]: done.  new sort implemented

**Comment by dtaylor113** [comment]
I believe I have addressed all review comments.  Pls see this PR's updated Description of screenshots of the usage table with unhuman labels and new sort.

File: frontend/public/components/resource-quota.jsx [hunk 22]
```diff
@@ -13,6 +13,18 @@ const menuActions = [...common];
 
 const quotaKind = quota => quota.metadata.namespace ? referenceForModel(ResourceQuotaModel) : referenceForModel(ClusterResourceQuotaModel);
 
+const compareResourceType = (left, right) => {
+  const regex = /\.|\//; //find object types delimited by '.' or '/'. Ex: '.cpu', '.memory', or '/persistentvolumeclaims'
+  left = left.match(regex) ? left.split(regex).reverse().join('') : left;
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]: You shouldn't need to check for matches. `split` will return an array with one value (the string) when there are no matches.

```suggestion
  left = left.split(regex).reverse().join('');
```

File: frontend/public/components/resource-quota.jsx [hunk 23]
```diff
@@ -13,6 +13,18 @@ const menuActions = [...common];
 
 const quotaKind = quota => quota.metadata.namespace ? referenceForModel(ResourceQuotaModel) : referenceForModel(ClusterResourceQuotaModel);
 
+const compareResourceType = (left, right) => {
+  const regex = /\.|\//; //find object types delimited by '.' or '/'. Ex: '.cpu', '.memory', or '/persistentvolumeclaims'
```
Comment (spadgett) [inline frontend/public/components/resource-quota.jsx]: To avoid repeating the split/reverse/join logic, I'd do something like

```js
// Reverse segments around `.` and `/` delimiters.
const reverseSegments = str => str.split(/\.|\//).reverse.join('');
left = reverseSegments(left);
right = reverseSegments(right);
return left.localCompare(right);
```

I wouldn't worry about case insensitive sorts if we're not sorting humanized values.

**Comment by spadgett** [comment]
I pushed a small fix for the last comment
